fem/kernel_dispatch.hpp

```diff
@@ -0,0 +1,281 @@
+// Declare the class used to dispatch shared memory kernels when the fallback methods don't
+// require template parameters.
+#define MFEM_DECLARE_KERNELS(KernelClassName, KernelType, UserParams, FallbackParams)        \
+class KernelClassName : public                                                               \
+   KernelsClassTemplate<KernelType,                                                          \
+   internal::KernelTypeList<UserParams>, internal::KernelTypeList<FallbackParams>>           \
+   {                                                                                         \
+   public:                                                                                   \
+      static KernelSignature Kernel1D();                                                     \
+      template<UserParams>                                                                   \
+      static KernelSignature Kernel2D();                                                     \
+      template<UserParams>                                                                   \
+      static KernelSignature Kernel3D();                                                     \
+      static KernelSignature Fallback2D();                                                   \
+      static KernelSignature Fallback3D();                                                   \
+   };
+
+constexpr int ipow(int x, int p) { return p == 0 ? 1 : x*ipow(x, p-1); }
+
+// This forward declaration of KernelClassTemplate allows a specialization accepting multiple variadic parameters.
+template<typename... T>
+class KernelsClassTemplate {};
+
+// KernelsClassTemplate holds the template methods for 2D and 3D kernels.  The class DispatchTable defined below
+// accesses `KernelClassTemplate` when assigning key value pairs, assigning a tuple key of `DIM, UserParams`
+// a specialized kernel.
+template<typename Signature, typename... UserParams, typename... FallbackParams>
+class KernelsClassTemplate<Signature, internal::KernelTypeList<UserParams...>, internal::KernelTypeList<FallbackParams...>>
+{
+private:
+   constexpr static int D(int D1D) { return (11 - D1D) / 2; }
+public:
+   using KernelSignature = Signature;
+
+   constexpr static int NBZ(int D1D, int Q1D)
+   {
+      return ipow(2, D(D1D) >= 0 ? D(D1D) : 0);
+   }
+
+   static KernelSignature Kernel1D();
+
+   template<UserParams... params>
+   static KernelSignature Kernel2D();
+
+   template<UserParams... params>
+   static KernelSignature Kernel3D();
+
+   template<FallbackParams... params>
+   static KernelSignature Fallback2D();
+
+   template<FallbackParams... params>
+   static KernelSignature Fallback3D();
+};
+
+// DispatchTable is the class that stores specialized kernels, assigning a key tuple of `KeyType`
+// a value of type `KernelType`.
+template <typename KeyType, typename KernelType, typename HashFunction>
+class DispatchTable
+{
+protected:
+   std::unordered_map<KeyType, KernelType, HashFunction> table;
+};
+
```

Comment:
Why not directly create the type and avoid the member function https://godbolt.org/z/hxv583o8z? This is unnecessary indirection. It avoids the `this->table.end()` calls and let's you work on an inherited version of `unordered_map`.